File: Decorator/Decorator.go
package main

import "fmt"

type Shape interface {
	draw()
}

type Circle struct {

}
func(c *Circle)draw(){
	fmt.Println("shape:Circle")
}


type Rectangle struct {

}

func(r *Rectangle)draw(){
	fmt.Println("shape:Rectangle")
}

type ShapeDector struct {
	dectorShape Shape
}

func(s *ShapeDector)draw(){
	s.dectorShape.draw()
}

type RedShapeDector struct {
	*ShapeDector
}
func(r *RedShapeDector)setRedBorder(shape Shape){
   fmt.Println("Bolder Color:Red")
}

func(r *RedShapeDector)draw(){
	r.ShapeDector.draw()
	r.setRedBorder(r.dectorShape)
}

func main(){
	circle:=new(Circle)
	redCircle:=&RedShapeDector{new(Circle)}
	redRectangle:=&RedShapeDector{new(Rectangle)}

	circle.draw()
	redCircle.draw()
	redRectangle.draw()
}

File: Decorator/Decorator.js
/**
 * Created by Jmesyan on 2017/3/23.
 */

/**
 interface Shape{
    public function draw();
}
 **/

var Circle=function(){
    this.draw=function(){
        console.log("Shape:Circle");
    }
};

var Rectangle=function(){
    this.draw=function(){
        console.log("Shape:Rectangle");
    }
};

var ShapeDector=function(dectorShape) {
    this.dectorShape = dectorShape;
};

ShapeDector.prototype.draw=function() {
   this.dectorShape.draw();
};

var RedShapeDector=function(dectorShape){
    ShapeDector.call(this,dectorShape);
}

RedShapeDector.prototype.draw=function() {
    ShapeDector.draw.call(this)

};


